eth/beacon/state_machines/base.py

```diff
@@ -193,9 +224,272 @@ def active_state(self) -> ActiveState:
+        processed_crystallized_state, processed_active_state = cls.compute_cycle_transitions(
+            crystallized_state,
+            processing_active_state,
+        )
+
+        # Return the copy
+        result_block = block.copy()
+        return result_block, processed_crystallized_state, processed_active_state
+
+    @classmethod
+    def compute_per_block_transtion(cls,
+                                    crystallized_state: CrystallizedState,
+                                    active_state: ActiveState,
+                                    block: BaseBeaconBlock,
+                                    chaindb: BaseBeaconChainDB,
+                                    cycle_length: int) -> ActiveState:
+        """
+        Process ``block`` and return the new ActiveState.
+
+
+        TODO: It doesn't match the latest spec.
+        There will be more fields need to be updated in ActiveState.
+        """
+        parent_block = chaindb.get_block_by_hash(block.parent_hash)
+        recent_block_hashes = get_new_recent_block_hashes(
+            active_state.recent_block_hashes,
+            parent_block.slot_number,
+            block.slot_number,
+            block.parent_hash
+        )
+
+        if parent_block.parent_hash != GENESIS_PARENT_HASH:
+            validate_parent_block_proposer(
+                crystallized_state,
+                block,
+                parent_block,
+                cycle_length,
+            )
+
+        # TODO: to implement the RANDAO reveal validation.
+        cls.validate_randao_reveal()
+        for attestation in block.attestations:
+            validate_attestation(
+                block,
+                parent_block,
+                crystallized_state,
+                recent_block_hashes,
+                attestation,
+                chaindb,
+                cycle_length,
+            )
+
+        return active_state.copy(
+            recent_block_hashes=recent_block_hashes,
+            pending_attestations=(
+                active_state.pending_attestations + block.attestations
+            ),
+        )
+
+    @classmethod
+    def compute_cycle_transitions(
+            cls,
+            crystallized_state: CrystallizedState,
+            active_state: ActiveState) -> Tuple[CrystallizedState, ActiveState]:
+        # TODO: it's a stub
+        return crystallized_state, active_state
+
+    #
+    #
+    # Proposer APIs
+    #
+    #
+    def propose_block(
+        self,
+        crystallized_state: CrystallizedState,
+        active_state: ActiveState,
+        block_proposal: 'BlockProposal',
+        chaindb: BaseBeaconChainDB,
+        config: BeaconConfig,
+        private_key: int
+    ) -> Tuple[BaseBeaconBlock, CrystallizedState, ActiveState, 'AttestationRecord']:
+        """
+        Propose the given block.
+        """
+        block, post_crystallized_state, post_active_state = self.process_block(
```

Comment:
If `process_block` is to be used for calculating the state roots for a block proposal, then it needs to have the option of being run without processing the signatures from the attestations. maybe another arg -- `validate_signatures`

We can assume the attestations being added were already checked as valid when initially received/created. No need to recheck here!